keps/sig-storage/20190204-kubernetes-csi-release-process.md

```diff
@@ -0,0 +1,302 @@
+- automatically generate release notes: this will be covered by a separate, future KEP
+
+
+## Proposal
+
+### Versionioning
+
+Each of the components has its own documentation, versioning and
+release notes. [Semantic versioning](https://semver.org/) with `v`
+prefix (i.e. `v1.2.3`) is used for components that provide a stable
+API, like for example:
+- external-attacher
+- external-provisioner
+- external-snapshotter
+- cluster-driver-registrar
+- node-driver-registrar
+- csi-driver-host-path
+- csi-lib-utils
+
+Other components are internal utility packages that are
+not getting tagged:
+- csi-release-tools
+
+An additional suffix like `-rc1` can be used to denote a pre-release.
+
+The [hostpath example
+deployment](https://github.com/kubernetes-csi/csi-driver-host-path/tree/master/deploy)
+defines the exact release of each sidecar that is part of that
+deployment. When updating to newer sidecar releases, a new hostpath
+driver release is tagged with version numbers bumped according to the
+same rules as the individual components (major version bump when any
+of the sidecars had a major change, etc.).
+
+#### Release artifacts
+
+Tagging a component with a `v*` tag triggers a
+build for that component with that version. The output is primarily the container image
+for components that need to be deployed. Those images get published as
+`gcr.io/kubernetes-csi/<image>:<tag>`, which depends on [issue
+158](https://github.com/kubernetes/k8s.io/issues/158) getting resolved.
+
+Only binaries provided as part of a release with a semantic version tag without additional suffix
+should be considered
+production ready. Binaries are never going to be rebuilt, therefore an
+image like `csi-node-driver-registrar:v1.0.2` will always pull the
+same content and `imagePullPolicy: Always` can be omitted.
+
+Eventually auxiliary files like the RBAC rules that are included in
+each component might also get published in a single location, like
+`https://dl.k8s.io/kubernetes-csi/`, but this currently outside of
+this KEP.
+
+#### Release process
+
+1. A change is submitted against the master branch.
+
+1. A [Prow](https://github.com/kubernetes/test-infra/blob/master/prow/README.md)
+   job checks out the source of the modified component, rebuilds it and then
+   runs unit tests and E2E tests with it as defined below.
+
+1. Maintainers accept changes into the master branch.
+
+1. The same Prow job runs for master and repeats the checks. If it succeeds,
+   a new "canary" image is published for the component.
+
+1. In preparation for the release of a major new update, a feature freeze is
+   declared for the "master" branch and only changes relevant for that next
+   release are accepted.
+
+1. When all changes targeted for the release are in master, automatic
+   test results are okay and and potentially some more manual tests,
+   maintainers tag a new release directly on the master branch. This
+   can be a `-rc` test release or a normal release.
```

Comment:
When is the `release-X.Y` branch cut?

I would propose:

1. A step before this where maintainer cuts the release `release-X.Y`
2. The verification in this test be done against the newly cut `release-X.Y` branch.
3. Instead of tagging a release from master branch, the maintainers tag release from the `release-X.Y` branch.

Reply:
A `release-X.Y` is cut when needed, which is the case when:
- a `vX.Y.Z` is to be released and
- `vX.Y.Z` cannot be tagged on the master branch because that already contains  changes that are not meant to be in that release.

I know that you prefer to create a release branch early, but that creates extra work for everyone (everything that is a bug fix needs to be backported by a dev and approved by a maintainer) and if a component only gets bug fixes, then master and maintenance branch will be identical in the end except that they have a different commit history.